MudDataGrid: Add Order parameter for explicit column ordering

janeswingler opened the pull request:
<!-- MAKE SURE TO READ THE CONTRIBUTING GUIDE: https://github.com/MudBlazor/MudBlazor/blob/dev/CONTRIBUTING.md -->
<!-- Describe your changes and what the purpose of them is. -->
<!-- If you made any visual changes, provide screenshots of before/after. If it has moving parts, please attach a high-quality video. -->

## Description

Closes [#10697](https://github.com/MudBlazor/MudBlazor/issues/10697)  
A revised implementation of the approach discussed in [#11918](https://github.com/MudBlazor/MudBlazor/pull/11918) at https://github.com/MudBlazor/MudBlazor/pull/11918#issuecomment-3347379575

Adds an `Order` parameter to all column types in `MudDataGrid` for controlling column display order. Enables repositioning of special columns (`Hierarchy`, `Select`) when needed.

#### Default behavior (unchanged):
- `HierarchyColumn` is inserted in `RenderedColumns` at index 0 → appears first  
- `SelectColumn` is inserted in `RenderedColumns` at index 1 if a `HierarchyColumn` is present, otherwise at index 0 → appears second, otherwise first  
- All other columns are added in markup order → appear after special columns, preserving declared markup order

#### With explicit `Order` values:
Columns are added as usual. When any column has `Order` set, a sort is applied via `SortColumnsByOrder()` where:
- Columns with `Order` → sorted by their `Order` value
- Columns without `Order` → use fallback priority (hierarchy=0, select=1, others=2)
- Lower values appear earlier

## Drag-and-Drop Interaction
Drag-and-drop nulls the `Order` parameter of both swapped columns. The rationale being that when a user drags a column, they're explicitly overriding declarative positioning. It gets tricky to infer user intent when dragging between columns with mixed Order states, e.g what value should a column get when dropped next to unordered columns? Also, rather than auto-assigning Order to all columns and then swapping on DragAndDrop, just nulling keeps Order opt-in.

## Key Changes

**Column.razor.cs:**  
- Added `Order` parameter (nullable int) with `ParameterState<int?>` pattern  
- Added `OrderChanged` event callback
- Added `OnOrderChangedAsync()` handler that triggers column re-sorting via `SortColumnsByOrder()`
- Updated constructor to register parameter state with change handler  

**MudDataGrid.razor.cs:**  
- Added `SortColumnsByOrder()` method to sort `RenderedColumns` in-place when any column has an explicit `Order` value
- Modified `ItemUpdatedAsync()` to null `Order` values of both columns involved in drag-and-drop operations

**HierarchyColumn.razor.cs / SelectColumn.razor.cs:**  
- Added `Order` parameter and `OrderChanged` event callback

**HierarchyColumn.razor / SelectColumn.razor:**  
- Pass `Order` and `OrderChanged` parameters through to internal `TemplateColumn`

**DataGridTests.cs:**  
- Added `DataGridColumnOrderTest()` - verifies `Order` parameter sorting behavior
- Added `DataGridColumnDragAndDropNullsOrderTest()` - verifies drag-and-drop nulls `Order` values

**Test Components:**  
- Created `DataGridColumnOrderTest.razor`
- Created `DataGridColumnDragAndDropNullsOrderTest.razor`

## Order Demo

The example below shows a `MudDataGrid` where columns are declared in this order:  
`Product`, `Order Date`, `Order #`, `Quantity`, `Total`, `SelectColumn`, `HierarchyColumn`.


https://github.com/user-attachments/assets/20dcb52e-0e0a-4b42-836c-e809d3d7762e

#### Initial State:  
`HierarchyColumn` appears first (`Order=0`), `SelectColumn` second (`Order=1`), and others follow (`Order=2`).

**Step 1:** Setting `Order=3` on `HierarchyColumn` moves it to the last position.  
**Step 2:** Setting `Order=0` on `Order #` moves it ahead of `SelectColumn`.  
**Step 3:** Setting `Order=-1` on `SelectColumn` moves it back to the first position.

## Order & DragAndDrop Demo
Same setup as above with DragDropColumnReordering enabled.
Set Order=3 on HierarchyColumn, then drag columns. Dragging columns overrides declared Order.

https://github.com/user-attachments/assets/3fb52212-0cdf-48d3-ae89-42fbce1600b4

## How Has This Been Tested
- [x] Added unit tests for `Order` parameter sorting behavior
- [x] Added unit tests for drag-and-drop Order nulling
- [x] Verified all existing DataGrid tests pass  
- [x] Visual testing in WASM and BSS  
- [x] Tested integration with local TryMudBlazor pointing to local MudBlazor solution  

## Type of Changes

- [ ] Bug fix (non-breaking change which fixes an issue)  
- [x] New feature (non-breaking change which adds functionality)  
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)  
- [ ] Documentation (fix or improvement to the website or code docs)  

## Checklist

- [x] The PR is submitted to the correct branch (`dev`).  
- [x] My code follows the code style of this project.  
- [x] I've added relevant tests and confirmed existing ones.

ScarletKuro commented:
@versile2, your quote
> In MudDataGrid.razor.cs adjust the, make an internal List SortedByColumns => RenderedColumns.OrderBy(x => x is HierarchyColumn ? 0 : x is SelectColumn ? 1 : 2).ThenBy(x => x.SortBy);`

Just wanted to ask if you think it's a good idea, since this introduces a breaking change. I know that many people use `RenderedColumns` directly, and I would also expect it to have the same order as the rendered output.

Perhaps we could manipulate the `RenderedColumns` order directly? I know it's generally harder to get right, but we could add an internal extension method like this:
```C#
void MoveItem<T>(List<T> list, int oldIndex, int newIndex)
{
    if (list == null || list.Count == 0)
        return;

    // Ignore invalid oldIndex
    if (oldIndex < 0 || oldIndex >= list.Count)
        return;

    // Clamp newIndex into valid range
    if (newIndex < 0)
        newIndex = 0;
    else if (newIndex >= list.Count)
        newIndex = list.Count - 1;

    // If same index, no move needed
    if (oldIndex == newIndex)
        return;

    T item = list[oldIndex];
    list.RemoveAt(oldIndex);

    // Adjust if we removed before the target position
    if (newIndex > oldIndex)
        newIndex--;

    list.Insert(newIndex, item);
}
```
```diff
-private void OnOrderChangedAsync()
-{
-    DataGrid?.DropContainerHasChanged();
-    ((IMudStateHasChanged)DataGrid)?.StateHasChanged();
-}

+private void OnOrderChangedAsync(ParameterChangedEventArgs<bool> args)
+{
+    DataGrid?.RenderedColumns.MoveItem(args.LastValue, args.Value);
+    DataGrid?.DropContainerHasChanged();
+    ((IMudStateHasChanged)DataGrid)?.StateHasChanged();
+}
```
Then, technically it wouldn't be a breaking change? Opinions?

versile2 commented:
> Just wanted to ask if you think it's a good idea, since this introduces a breaking change. I know that many people use `RenderedColumns` directly, and I would also expect it to have the same order as the rendered output.

Well if they do use the Rendered Columns it will still sort the exact same way unless they fill out the orderby field. I would prefer to make the change in Rendered columns, and I guess we probably could just by sorting it if there is an order by?

ScarletKuro commented:
> @janeswingler Any thoughts on the above conversations? It would be nice to get this in before December if possible.

We can merge it in v9 to be sure?

janeswingler commented:
> @janeswingler Any thoughts on the above conversations? It would be nice to get this in before December if possible.

Sounds good

janeswingler committed:
DataGrid: Add Order parameter to control column display order on special columns (#10697)

- Add Order parameter to Column, HierarchyColumn, and SelectColumn
- Implement SortColumnsByOrder() method for column reordering
- Null Order values on drag-and-drop to allow manual override
- Add unit tests for Order sorting and drag-and-drop behavior

versile2 commented on src/MudBlazor/Components/DataGrid/MudDataGrid.razor.cs:
```diff
@@ -234,6 +234,10 @@ private Task ItemUpdatedAsync(MudItemDropInfo<Column<T>> dropItem)
 
+                // Null both Order values on drag-and-drop
+                dragAndDropSource.OrderState.SetValueAsync(null).CatchAndLog();
+                dragAndDropDestination.OrderState.SetValueAsync(null).CatchAndLog();
```
I think here you should know the order of the column? And on drag and drop report the order to the Order value if it's not null already? Part of the big reason for this is so users can track the order of columns and start it however their customer left it off. I don't think they can do that with order being set to null on drag / drop?

ScarletKuro replied:
Why it's `CatchAndLog`? we never use that if the method is task, pls await it.
